R/convenience-functions.R

```diff
@@ -0,0 +1,194 @@
+  scale_col_present <- ("scale" %in% colnames(original_data))
+
+  # Error handling
+  if (scale_col_present) {
+    if (!("natural" %in% original_data$scale)) {
+      stop("If a column 'scale' is present, entries with scale =='natural' are required for the transformation")
+    }
+    if (append && (label %in% original_data$scale)) {
+      w <- paste0(
+        "Appending new transformations with label '", label,
+        "', even though that entry is already present in column 'scale'."
+      )
+      warning(w)
+    }
+  }
+
+  if (append) {
+    if (scale_col_present) {
+      transformed_data <- data.table::copy(original_data)[scale == "natural"]
+    } else {
+      transformed_data <- data.table::copy(original_data)
+      original_data[, scale := "natural"]
+    }
+    transformed_data[, prediction := fun(prediction, ...)]
+    transformed_data[, true_value := fun(true_value, ...)]
+    transformed_data[, scale := label]
+    out <- rbind(original_data, transformed_data)
+    return(out[])
+  }
+
+  # check if a column called "scale" is already present and if so, only
+  # restrict to transformations of the original data
+  if (scale_col_present) {
+    original_data[scale == "natural", prediction := fun(prediction, ...)]
+    original_data[scale == "natural", true_value := fun(true_value, ...)]
+    original_data[scale == "natural", scale := label]
+  } else {
+    original_data[, prediction := fun(prediction, ...)]
+    original_data[, true_value := fun(true_value, ...)]
+  }
+  return(original_data[])
+}
+
+
+
+
+
+
+#' @title Log transformation with an additive shift
+#'
+#' @description Function that shifts a value by some offset and then applies the
+#' natural logarithm to it.
+#'
+#' @details The output is computed as log(x + offset) (or
+#' log(pmax(0, x) + offset)) if `truncate = TRUE`.
+#'
+#' @param x vector of input values to be transformed
+#' @param offset number to add to the input value before taking the natural
+#' logarithm
+#' @param truncate whether to truncate negative values to 0. Default is FALSE.
+#' @param base a positive or complex number: the base with respect to which
+#' logarithms are computed. Defaults to e = exp(1).
+#' @return A numeric vector with transformed values
+#' @export
+#' @references Transformation of forecasts for evaluating predictive
+#' performance in an epidemiological context
+#' Nikos I. Bosse, Sam Abbott, Anne Cori, Edwin van Leeuwen, Johannes Bracher,
+#' Sebastian Funk
+#' medRxiv 2023.01.23.23284722
+#' \doi{https://doi.org/10.1101/2023.01.23.23284722}
+#' <https://www.medrxiv.org/content/10.1101/2023.01.23.23284722v1> # nolint
+
+#' @keywords check-forecasts
+#' @examples
+#'
+#' log_shift(1:10)
+#' log_shift(0:9, offset = 1)
+#' log_shift(-1:9, offset = 1, truncate = TRUE)
+#'
+#' transform_forecasts(
+#'   example_quantile,
+#'   fun = log_shift,
+#'   offset = 1,
+#'   truncate = TRUE
+#'  )
+
+log_shift <- function(
+    x,
+    offset = 0,
+    truncate = FALSE,
```

Comment:
Do we really want this option here? I'd think it's better for the log function to throw an error and the user to pre-process the forecasts/data so that there are no negatives.

If sticking with it I would suggest to make this more self-explanatory, i.e. calling the option `negative_to_zero` or something like that (although given that we might be adding an offset anyway I'm not clear why we'd choose zero over anything else).

Reply:
I think the reason to have that truncation is that the alternative (at least staying in the same workflow) is a bit more clunky. You would have to use something like `transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural")`

See e.g. these examples: 

```
#' example_quantile %>%
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") 
```

```
#' # adding multiple transformations
#' library(magrittr) # pipe operator
#' example_quantile %>%
#'   transform_forecasts(offset = 1, truncate = TRUE) %>%
#'   # manually truncate all negative values before applying sqrt
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural") %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
#'   score() %>%
#'   summarise_scores(by = c("model", "scale"))
```

(the part with the label = natural here is only because it's transforming the original while something else has already been computed, which maybe is a stupid thing to do?)

Reply:
Looking at it again I should definitely simplify the second example...

Reply:
> I think the reason to have that truncation is that the alternative (at least staying in the same workflow) is a bit more clunky. You would have to use something like `transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural")`

Wouldn't it be more natural to do something like:

```r
example_quantile %>%
  dplyr::filter(true_value >= 0) %>%
  transform_forecasts(fun = sqrt, label = "sqrt") 
```

or 

```r
example_quantile %>%
  dplyr::mutate(true_value = dplyr::ifelse(true_value < 0, 0, true_value)) %>%
  transform_forecasts(fun = sqrt, label = "sqrt") 
```

i.e. fix the problem in the data rather than letting the transformation function take care of it?